fact/src/cgroup.rs

```diff
@@ -0,0 +1,174 @@
+use std::{
+    collections::HashMap,
+    os::unix::fs::DirEntryExt,
+    path::PathBuf,
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
+
+use log::warn;
+use tokio::{
+    sync::{watch::Receiver, Mutex},
+    task::JoinHandle,
+    time,
+};
+
+use crate::host_info::get_cgroup_paths;
+
+#[derive(Debug)]
+struct ContainerIdEntry {
+    container_id: Option<Arc<String>>,
+    pub last_seen: SystemTime,
+}
+
+type ContainerIdMap = HashMap<u64, ContainerIdEntry>;
+
+#[derive(Debug, Clone, Default)]
+pub struct ContainerIdCache(Arc<Mutex<ContainerIdMap>>);
+
+impl ContainerIdCache {
+    pub fn new() -> Self {
+        let mut map = HashMap::new();
+        ContainerIdCache::update_unlocked(&mut map);
+        ContainerIdCache(Arc::new(Mutex::new(map)))
+    }
+
+    fn update_unlocked(map: &mut ContainerIdMap) {
+        for root in get_cgroup_paths() {
+            ContainerIdCache::walk_cgroupfs(&root, map, None);
+        }
+    }
+
+    async fn update(&mut self) {
+        let mut map = self.0.lock().await;
+        ContainerIdCache::update_unlocked(&mut map);
+    }
+
+    async fn prune(&mut self) {
+        let now = SystemTime::now();
+        self.0.lock().await.retain(|_, value| {
+            now.duration_since(value.last_seen).unwrap() < Duration::from_secs(30)
```

Comment:
Maybe set a variable to 30 seconds and maybe pass that variable to this method to make this more configurable. Also maybe change the name to something like `prun_old`.